File: java/analyze/PipelineStage.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.typedb.driver.analyze;

import com.typedb.driver.common.NativeIterator;
import com.typedb.driver.common.NativeObject;
import com.typedb.driver.common.exception.TypeDBDriverException;

import java.util.stream.Stream;

import static com.typedb.driver.common.exception.ErrorMessage.Analyze.INVALID_CONSTRAINT_CASTING;
import static com.typedb.driver.common.exception.ErrorMessage.Analyze.INVALID_STAGE_CASTING;
import static com.typedb.driver.common.exception.ErrorMessage.Internal.UNEXPECTED_NATIVE_VALUE;
import static com.typedb.driver.common.util.Objects.className;

public abstract class PipelineStage extends NativeObject<com.typedb.driver.jni.PipelineStage> {

    PipelineStage(com.typedb.driver.jni.PipelineStage nativeObject) {
        super(nativeObject);
    }


    public com.typedb.driver.jni.PipelineStageVariant getVariant() {
        return com.typedb.driver.jni.typedb_driver.pipeline_stage_variant(nativeObject);
    }

    public boolean isMatch() {
        return false;
    }

    public boolean isInsert() {
        return false;
    }

    public boolean isPut() {
        return false;
    }

    public boolean isUpdate() {
        return false;
    }

    public boolean isDelete() {
        return false;
    }

    public boolean isSelect() {
        return false;
    }

    public boolean isSort() {
        return false;
    }

    public boolean isRequire() {
        return false;
    }

    public boolean isOffset() {
        return false;
    }

    public boolean isLimit() {
        return false;
    }

    public boolean isDistinct() {
        return false;
    }

    public boolean isReduce() {
        return false;
    }


    public MatchStage asMatch() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(MatchStage.class));
    }

    public InsertStage asInsert() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(InsertStage.class));
    }

    public PutStage asPut() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(PutStage.class));
    }

    public UpdateStage asUpdate() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(UpdateStage.class));
    }

    public DeleteStage asDelete() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(DeleteStage.class));
    }

    public SelectStage asSelect() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(SelectStage.class));
    }

    public SortStage asSort() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(SortStage.class));
    }

    public RequireStage asRequire() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(RequireStage.class));
    }

    public OffsetStage asOffset() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(OffsetStage.class));
    }

    public LimitStage asLimit() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(LimitStage.class));
    }

    public DistinctStage asDistinct() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(DistinctStage.class));
    }

    public ReduceStage asReduce() {
        throw new TypeDBDriverException(INVALID_STAGE_CASTING, className(this.getClass()), className(ReduceStage.class));
    }

    public static PipelineStage of(com.typedb.driver.jni.PipelineStage nativeObject) {
        switch (com.typedb.driver.jni.typedb_driver.pipeline_stage_variant(nativeObject)) {
            case Match:
                return new MatchStage(nativeObject);
            case Insert:
                return new InsertStage(nativeObject);
            case Put:
                return new PutStage(nativeObject);
            case Update:
                return new UpdateStage(nativeObject);
            case Delete:
                return new DeleteStage(nativeObject);
            case Select:
                return new SelectStage(nativeObject);
            case Sort:
                return new SortStage(nativeObject);
            case Require:
                return new RequireStage(nativeObject);
            case Offset:
                return new OffsetStage(nativeObject);
            case Limit:
                return new LimitStage(nativeObject);
            case Distinct:
                return new DistinctStage(nativeObject);
            case Reduce:
                return new ReduceStage(nativeObject);
            default:
                throw new TypeDBDriverException(UNEXPECTED_NATIVE_VALUE);
        }
    }

    public static class MatchStage extends PipelineStage {

        MatchStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isMatch() {
            return true;
        }

        @Override
        public MatchStage asMatch() {
            return this;
        }

        public ConjunctionID block() {
            return new ConjunctionID(com.typedb.driver.jni.typedb_driver.pipeline_stage_get_block(nativeObject));
        }
    }

    public static class InsertStage extends PipelineStage {

        InsertStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isInsert() {
            return true;
        }

        @Override
        public InsertStage asInsert() {
            return this;
        }

        public ConjunctionID block() {
            return new ConjunctionID(com.typedb.driver.jni.typedb_driver.pipeline_stage_get_block(nativeObject));
        }
    }

    public static class PutStage extends PipelineStage {
        PutStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isPut() {
            return true;
        }

        @Override
        public PutStage asPut() {
            return this;
        }

        public ConjunctionID block() {
            return new ConjunctionID(com.typedb.driver.jni.typedb_driver.pipeline_stage_get_block(nativeObject));
        }
    }

    public static class UpdateStage extends PipelineStage {
        UpdateStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isUpdate() {
            return true;
        }

        @Override
        public UpdateStage asUpdate() {
            return this;
        }

        public ConjunctionID block() {
            return new ConjunctionID(com.typedb.driver.jni.typedb_driver.pipeline_stage_get_block(nativeObject));
        }
    }

    public static class DeleteStage extends PipelineStage {
        DeleteStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isDelete() {
            return true;
        }

        @Override
        public DeleteStage asDelete() {
            return this;
        }

        public ConjunctionID block() {
            return new ConjunctionID(com.typedb.driver.jni.typedb_driver.pipeline_stage_get_block(nativeObject));
        }

        public Stream<Variable> deletedVariables() {
            return new NativeIterator<>(com.typedb.driver.jni.typedb_driver.pipeline_stage_delete_get_deleted_variables(nativeObject)).stream().map(Variable::new);
        }
    }

    public static class SelectStage extends PipelineStage {
        SelectStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isSelect() {
            return true;
        }

        @Override
        public SelectStage asSelect() {
            return this;
        }


        public Stream<Variable> variables() {
            return new NativeIterator<com.typedb.driver.jni.Variable>(com.typedb.driver.jni.typedb_driver.pipeline_stage_select_get_variables(nativeObject)).stream().map(Variable::new);
        }
    }

    public static class SortStage extends PipelineStage {
        SortStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isSort() {
            return true;
        }

        @Override
        public SortStage asSort() {
            return this;
        }

        public Stream<SortVariable> variables() {
            return new NativeIterator<com.typedb.driver.jni.SortVariable>(
                    com.typedb.driver.jni.typedb_driver.pipeline_stage_sort_get_sort_variables(nativeObject)
            ).stream().map(SortVariable::new);
        }

        public static class SortVariable extends NativeObject<com.typedb.driver.jni.SortVariable> {

            protected SortVariable(com.typedb.driver.jni.SortVariable nativeObject) {
                super(nativeObject);
            }

            public Variable variable() {
                return new Variable(com.typedb.driver.jni.typedb_driver.sort_variable_get_variable(nativeObject));
            }

            public com.typedb.driver.jni.SortOrder order() {
                return com.typedb.driver.jni.typedb_driver.sort_variable_get_order(nativeObject);
            }
        }
    }

    public static class RequireStage extends PipelineStage {
        RequireStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isRequire() {
            return true;
        }

        @Override
        public RequireStage asRequire() {
            return this;
        }

        public Stream<Variable> variables() {
            return new NativeIterator<com.typedb.driver.jni.Variable>(com.typedb.driver.jni.typedb_driver.pipeline_stage_require_get_variables(nativeObject)).stream().map(Variable::new);
        }
    }

    public static class OffsetStage extends PipelineStage {
        OffsetStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isOffset() {
            return true;
        }

        @Override
        public OffsetStage asOffset() {
            return this;
        }

        public long offset() {
            return com.typedb.driver.jni.typedb_driver.pipeline_stage_offset_get_offset(nativeObject);
        }
    }

    public static class LimitStage extends PipelineStage {
        LimitStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isLimit() {
            return true;
        }

        @Override
        public LimitStage asLimit() {
            return this;
        }


        public long limit() {
            return com.typedb.driver.jni.typedb_driver.pipeline_stage_limit_get_limit(nativeObject);
        }
    }

    public static class DistinctStage extends PipelineStage {
        DistinctStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isDistinct() {
            return true;
        }

        @Override
        public DistinctStage asDistinct() {
            return this;
        }
    }

    public static class ReduceStage extends PipelineStage {
        ReduceStage(com.typedb.driver.jni.PipelineStage nativeObject) {
            super(nativeObject);
        }

        @Override
        public boolean isReduce() {
            return true;
        }

        @Override
        public ReduceStage asReduce() {
            return this;
        }

        public Stream<Variable> groupBy() {
            return new NativeIterator<com.typedb.driver.jni.Variable>(
                    com.typedb.driver.jni.typedb_driver.pipeline_stage_reduce_get_groupby(nativeObject)
            ).stream().map(Variable::new);
        }

        public Stream<ReduceAssignment> reducerAssignments() {
            return new NativeIterator<com.typedb.driver.jni.ReduceAssignment>(
                    com.typedb.driver.jni.typedb_driver.pipeline_stage_reduce_get_reducer_assignments(nativeObject)
            ).stream().map(ReduceAssignment::new);
        }

        public static class ReduceAssignment extends NativeObject<com.typedb.driver.jni.ReduceAssignment> {

            protected ReduceAssignment(com.typedb.driver.jni.ReduceAssignment nativeObject) {
                super(nativeObject);
            }

            public Variable assigned() {
                return new Variable(com.typedb.driver.jni.typedb_driver.reduce_assignment_get_assigned(nativeObject));
            }

            public Reducer reducer() {
                return new Reducer(com.typedb.driver.jni.typedb_driver.reduce_assignment_get_reducer(nativeObject));
            }
        }
    }
}

File: csharp/Analyze/PipelineStage.cs
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

using System;
using System.Collections.Generic;
using System.Linq;

using TypeDB.Driver.Api.Analyze;
using TypeDB.Driver.Common;

namespace TypeDB.Driver.Analyze
{
    public abstract class PipelineStage : NativeObjectWrapper<Pinvoke.PipelineStage>, IPipelineStage
    {
        protected PipelineStage(Pinvoke.PipelineStage nativeObject)
            : base(nativeObject)
        {
        }

        public static PipelineStage Of(Pinvoke.PipelineStage nativeObject)
        {
            var variant = Pinvoke.typedb_driver.pipeline_stage_variant(nativeObject);
            switch (variant)
            {
                case Pinvoke.PipelineStageVariant.Match:
                    return new MatchStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Insert:
                    return new InsertStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Put:
                    return new PutStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Update:
                    return new UpdateStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Delete:
                    return new DeleteStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Select:
                    return new SelectStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Sort:
                    return new SortStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Require:
                    return new RequireStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Offset:
                    return new OffsetStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Limit:
                    return new LimitStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Distinct:
                    return new DistinctStageImpl(nativeObject);
                case Pinvoke.PipelineStageVariant.Reduce:
                    return new ReduceStageImpl(nativeObject);
                default:
                    throw new InvalidOperationException("Unexpected pipeline stage variant: " + variant);
            }
        }

        public Pinvoke.PipelineStageVariant Variant
        {
            get { return Pinvoke.typedb_driver.pipeline_stage_variant(NativeObject); }
        }

        public virtual bool IsMatch => false;
        public virtual bool IsInsert => false;
        public virtual bool IsPut => false;
        public virtual bool IsUpdate => false;
        public virtual bool IsDelete => false;
        public virtual bool IsSelect => false;
        public virtual bool IsSort => false;
        public virtual bool IsRequire => false;
        public virtual bool IsOffset => false;
        public virtual bool IsLimit => false;
        public virtual bool IsDistinct => false;
        public virtual bool IsReduce => false;

        public virtual IMatchStage AsMatch() => throw InvalidCast("MatchStage");
        public virtual IInsertStage AsInsert() => throw InvalidCast("InsertStage");
        public virtual IPutStage AsPut() => throw InvalidCast("PutStage");
        public virtual IUpdateStage AsUpdate() => throw InvalidCast("UpdateStage");
        public virtual IDeleteStage AsDelete() => throw InvalidCast("DeleteStage");
        public virtual ISelectStage AsSelect() => throw InvalidCast("SelectStage");
        public virtual ISortStage AsSort() => throw InvalidCast("SortStage");
        public virtual IRequireStage AsRequire() => throw InvalidCast("RequireStage");
        public virtual IOffsetStage AsOffset() => throw InvalidCast("OffsetStage");
        public virtual ILimitStage AsLimit() => throw InvalidCast("LimitStage");
        public virtual IDistinctStage AsDistinct() => throw InvalidCast("DistinctStage");
        public virtual IReduceStage AsReduce() => throw InvalidCast("ReduceStage");

        private InvalidOperationException InvalidCast(string targetType)
        {
            return new InvalidOperationException($"Cannot cast {GetType().Name} to {targetType}");
        }

        public override string ToString()
        {
            return Pinvoke.typedb_driver.pipeline_stage_string_repr(NativeObject);
        }

        public class MatchStageImpl : PipelineStage, IMatchStage
        {
            internal MatchStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsMatch => true;
            public override IMatchStage AsMatch() => this;

            public IConjunctionID Block =>
                new ConjunctionID(Pinvoke.typedb_driver.pipeline_stage_get_block(NativeObject));
        }

        public class InsertStageImpl : PipelineStage, IInsertStage
        {
            internal InsertStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsInsert => true;
            public override IInsertStage AsInsert() => this;

            public IConjunctionID Block =>
                new ConjunctionID(Pinvoke.typedb_driver.pipeline_stage_get_block(NativeObject));
        }

        public class PutStageImpl : PipelineStage, IPutStage
        {
            internal PutStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsPut => true;
            public override IPutStage AsPut() => this;

            public IConjunctionID Block =>
                new ConjunctionID(Pinvoke.typedb_driver.pipeline_stage_get_block(NativeObject));
        }

        public class UpdateStageImpl : PipelineStage, IUpdateStage
        {
            internal UpdateStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsUpdate => true;
            public override IUpdateStage AsUpdate() => this;

            public IConjunctionID Block =>
                new ConjunctionID(Pinvoke.typedb_driver.pipeline_stage_get_block(NativeObject));
        }

        public class DeleteStageImpl : PipelineStage, IDeleteStage
        {
            internal DeleteStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsDelete => true;
            public override IDeleteStage AsDelete() => this;

            public IConjunctionID Block =>
                new ConjunctionID(Pinvoke.typedb_driver.pipeline_stage_get_block(NativeObject));

            public IEnumerable<IVariable> DeletedVariables =>
                new NativeEnumerable<Pinvoke.Variable>(
                    Pinvoke.typedb_driver.pipeline_stage_delete_get_deleted_variables(NativeObject))
                    .Select(v => new Variable(v));
        }

        public class SelectStageImpl : PipelineStage, ISelectStage
        {
            internal SelectStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsSelect => true;
            public override ISelectStage AsSelect() => this;

            public IEnumerable<IVariable> Variables =>
                new NativeEnumerable<Pinvoke.Variable>(
                    Pinvoke.typedb_driver.pipeline_stage_select_get_variables(NativeObject))
                    .Select(v => new Variable(v));
        }

        public class SortStageImpl : PipelineStage, ISortStage
        {
            internal SortStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsSort => true;
            public override ISortStage AsSort() => this;

            public IEnumerable<ISortVariable> Variables =>
                new NativeEnumerable<Pinvoke.SortVariable>(
                    Pinvoke.typedb_driver.pipeline_stage_sort_get_sort_variables(NativeObject))
                    .Select(sv => new SortVariableImpl(sv));
        }

        public class SortVariableImpl : NativeObjectWrapper<Pinvoke.SortVariable>, ISortVariable
        {
            internal SortVariableImpl(Pinvoke.SortVariable nativeObject)
                : base(nativeObject)
            {
            }

            public IVariable Variable =>
                new Variable(Pinvoke.typedb_driver.sort_variable_get_variable(NativeObject));

            public Pinvoke.SortOrder Order =>
                Pinvoke.typedb_driver.sort_variable_get_order(NativeObject);
        }

        public class RequireStageImpl : PipelineStage, IRequireStage
        {
            internal RequireStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsRequire => true;
            public override IRequireStage AsRequire() => this;

            public IEnumerable<IVariable> Variables =>
                new NativeEnumerable<Pinvoke.Variable>(
                    Pinvoke.typedb_driver.pipeline_stage_require_get_variables(NativeObject))
                    .Select(v => new Variable(v));
        }

        public class OffsetStageImpl : PipelineStage, IOffsetStage
        {
            internal OffsetStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsOffset => true;
            public override IOffsetStage AsOffset() => this;

            public long Offset => Pinvoke.typedb_driver.pipeline_stage_offset_get_offset(NativeObject);
        }

        public class LimitStageImpl : PipelineStage, ILimitStage
        {
            internal LimitStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsLimit => true;
            public override ILimitStage AsLimit() => this;

            public long Limit => Pinvoke.typedb_driver.pipeline_stage_limit_get_limit(NativeObject);
        }

        public class DistinctStageImpl : PipelineStage, IDistinctStage
        {
            internal DistinctStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsDistinct => true;
            public override IDistinctStage AsDistinct() => this;
        }

        public class ReduceStageImpl : PipelineStage, IReduceStage
        {
            internal ReduceStageImpl(Pinvoke.PipelineStage nativeObject) : base(nativeObject) { }

            public override bool IsReduce => true;
            public override IReduceStage AsReduce() => this;

            public IEnumerable<IVariable> GroupBy =>
                new NativeEnumerable<Pinvoke.Variable>(
                    Pinvoke.typedb_driver.pipeline_stage_reduce_get_groupby(NativeObject))
                    .Select(v => new Variable(v));

            public IEnumerable<IReduceAssignment> ReducerAssignments =>
                new NativeEnumerable<Pinvoke.ReduceAssignment>(
                    Pinvoke.typedb_driver.pipeline_stage_reduce_get_reducer_assignments(NativeObject))
                    .Select(ra => new ReduceAssignmentImpl(ra));
        }

        public class ReduceAssignmentImpl : NativeObjectWrapper<Pinvoke.ReduceAssignment>, IReduceAssignment
        {
            internal ReduceAssignmentImpl(Pinvoke.ReduceAssignment nativeObject)
                : base(nativeObject)
            {
            }

            public IVariable Assigned =>
                new Variable(Pinvoke.typedb_driver.reduce_assignment_get_assigned(NativeObject));

            public IReducer Reducer =>
                new Reducer(Pinvoke.typedb_driver.reduce_assignment_get_reducer(NativeObject));
        }
    }
}
